core/Pos/Crypto/HD.hs

```diff
@@ -0,0 +1,103 @@
+-- | Hierarchical derivation interface.
+
+module Pos.Crypto.HD
+       ( HDPassphrase
+       , HDAddressPayload (..)
+       , packHDAddressAttr
+       , deriveHDPublicKey
+       , deriveHDSecretKey
+       , deriveHDPassphrase
+       ) where
+
+import           Cardano.Crypto.Wallet        (deriveXPrv, deriveXPrvHardened, deriveXPub,
+                                               unXPub)
+import           Crypto.Cipher.ChaChaPoly1305 as C
+import           Crypto.Error
+import           Crypto.Hash                  (SHA512 (..))
+import qualified Crypto.KDF.PBKDF2            as PBKDF2
+import           Data.Binary.Put              (runPut)
+import           Data.ByteArray               as BA (ByteArrayAccess, convert)
+import           Data.ByteString.Char8        as B
+import qualified Data.ByteString.Lazy         as BSL
+import           Universum
+
+import           Pos.Binary.Class             (Bi (..))
+import           Pos.Crypto.Signing           (PublicKey (..), SecretKey (..))
+
+-- | Passphrase is a hash of root public key.
+-- We don't use root public key for store money, we use hash of it instead.
+newtype HDPassphrase = HDPassphrase ByteString
+
+-- | HDAddressPayload consists of
+-- * serialiazed and encrypted with symmetric scheme with passphrase (via ChaChaPoly1305 algorithm)
+-- * cryptographic tag
+-- For more information see 'packHDAddressAttr' and 'encryptChaChaPoly'.
+newtype HDAddressPayload = HDAddressPayload ByteString
+    deriving (Eq, Ord, Show, NFData, Generic)
+
+-- | Compute passphrase as hash of the root public key.
+deriveHDPassphrase :: PublicKey -> HDPassphrase
+deriveHDPassphrase (PublicKey pk) = HDPassphrase $
+    PBKDF2.generate
+        (PBKDF2.prfHMAC SHA512)
+        (PBKDF2.Parameters 500 passLen)
```

Comment:
@volhovM 
see here:
https://github.com/serokell/kademlia/blob/master/src/Network/Kademlia/HashNodeId.hs#L36
I'll add this comment too